### PR TITLE
fix decompress zstd error

### PR DESCRIPTION
**Pull request by fcharlie**
@nmoinvaz 
I tried it tonight(UTC+8), and there seems to be an error in your zstd decompression code. This is an improvement, but I did not add enough tests, I hope you can review it. Thanks.

### CODE REVIEW

File: mz_strm_zstd.c [hunk 1]
```diff
@@ -132,31 +129,21 @@ int32_t mz_stream_zstd_read(void *stream, void *buf, int32_t size) {
...
         total_in_before = zstd->in.pos;
-
         result = ZSTD_decompressStream(zstd->zdstream, &zstd->out, &zstd->in);
-
         if (ZSTD_isError(result)) {
             zstd->error = (int32_t)result;
             return MZ_DATA_ERROR;
         }
-
-        total_in_after = zstd->in.pos;
```
Comment (nmoinvaz) [inline mz_strm_zstd.c:147]: I would like to keep things the way they were with these extra variables. I realize that some of the variables are not necessary. But it keeps all the other streams the same way.

File: mz_strm_zstd.c [hunk 2]
```diff
@@ -132,31 +129,21 @@ int32_t mz_stream_zstd_read(void *stream, void *buf, int32_t size) {
             zstd->in.src = zstd->buffer;
             zstd->in.size = read;
             zstd->in.pos = 0;
+            zstd->total_in += read;
         }
 
         total_out_before = zstd->out.pos;
         total_in_before = zstd->in.pos;
-
```
Comment (nmoinvaz) [inline mz_strm_zstd.c:139]: Keep whitespaces.

**Comment by nmoinvaz** [comment]
Do you have the test file this failed on initially?

**Comment by fcharlie** [comment]
Test zip(zstd) file:
[bali-windows-amd64-1.0.0.zip](https://github.com/nmoinvaz/minizip/files/4783253/bali-windows-amd64-1.0.0.zip)

**Comment by gvollant** [comment]
For info, I had another idea for decompress loop

```
int32_t mz_stream_zstd_read(void *stream, void *buf, int32_t size)
{
#ifdef MZ_ZIP_NO_DECOMPRESSION
    MZ_UNUSED(stream);
    MZ_UNUSED(buf);
    MZ_UNUSED(size);
    return MZ_SUPPORT_ERROR;
#else
    mz_stream_zstd *mz_zstd = (mz_stream_zstd *)stream;
    uint64_t total_in_before = 0;
    uint64_t total_out_before = 0;
    uint64_t total_in_after = 0;
    uint64_t total_out_after = 0;
    int32_t total_in = 0;
    int32_t total_out = 0;
    int32_t in_bytes = 0;
    int32_t out_bytes = 0;
    int32_t bytes_to_read = 0;
    int32_t read = 0;
    size_t errorCode = 0;
    //int32_t err = LZMA_OK;
    mz_zstd->out.dst = (void*)buf;
    mz_zstd->out.size = (size_t)size;
    mz_zstd->out.pos = 0;
    for(;;)
    {
        if (mz_zstd->in.pos == mz_zstd->in.size)
        {
            bytes_to_read = sizeof(mz_zstd->buffer);
            if (mz_zstd->max_total_in > 0)
            {
                if ((int64_t)bytes_to_read > (mz_zstd->max_total_in - mz_zstd->total_in))
                    bytes_to_read = (int32_t)(mz_zstd->max_total_in - mz_zstd->total_in);
            }

            read = mz_stream_read(mz_zstd->stream.base, mz_zstd->buffer, bytes_to_read);

            if (read < 0)
                return read;

            mz_zstd->in.src = (const void*)mz_zstd->buffer;
            mz_zstd->in.pos = 0;
            mz_zstd->in.size = (size_t)read;
        }

        total_in_before = mz_zstd->in.pos;
        total_out_before = mz_zstd->out.pos;

        errorCode = ZSTD_decompressStream(mz_zstd->dstream, &mz_zstd->out, &mz_zstd->in);

        if (ZSTD_isError(errorCode))
        {
            mz_zstd->error = errorCode;
            break;
        }
        
        total_in_after = mz_zstd->in.pos;
        total_out_after = mz_zstd->out.pos;
        if ((mz_zstd->max_total_out != -1) && (int64_t)total_out_after > mz_zstd->max_total_out)
            total_out_after = (uint64_t)mz_zstd->max_total_out;

        in_bytes = (int32_t)(total_in_before - total_in_after);
        out_bytes = (int32_t)(total_out_after - total_out_before);

        total_in += in_bytes;
        total_out += out_bytes;

        mz_zstd->total_in += in_bytes;
        mz_zstd->total_out += out_bytes;

        if (mz_zstd->in.pos == mz_zstd->in.size && mz_zstd->out.pos < mz_zstd->out.size) 
            break;  
    }
    

    if (ZSTD_isError(errorCode))
        return MZ_DATA_ERROR;

    return total_out;
#endif
}
```

**Comment by fcharlie** [comment]
@gvollant Your suggestion is good, but there are a few minor mistakes, and I revised it. Thanks.

**Comment by nmoinvaz** [comment]
It seems like CI instances are failing..

File: mz_strm_zstd.c [hunk 3]
```diff
@@ -104,59 +104,71 @@ int32_t mz_stream_zstd_read(void *stream, void *buf, int32_t size) {
...
+    zstd->out.dst = (void*)buf;
+    zstd->out.size = (size_t)size;
     zstd->out.pos = 0;
-    zstd->out.dst = buf;
-    zstd->out.size = size;
 
-    do {
-        if (zstd->in.pos == zstd->in.size) {
-            if (zstd->max_total_in > 0) {
+    for(;;)
```
Comment (nmoinvaz) [inline mz_strm_zstd.c:122]: Change back to `do` loop please?

File: mz_strm_zstd.c [hunk 4]
```diff
@@ -104,59 +104,71 @@ int32_t mz_stream_zstd_read(void *stream, void *buf, int32_t size) {
...
             zstd->in.pos = 0;
+            zstd->in.size = (size_t)read;
         }
 
-        total_out_before = zstd->out.pos;
         total_in_before = zstd->in.pos;
+        total_out_before = zstd->out.pos;
 
-        result = ZSTD_decompressStream(zstd->zdstream, &zstd->out, &zstd->in);
+        errorCode = ZSTD_decompressStream(zstd->zdstream, &zstd->out, &zstd->in);
```
Comment (nmoinvaz) [inline mz_strm_zstd.c:146]: Code styling should be `error_code`.

File: mz_strm_zstd.c [hunk 5]
```diff
@@ -104,59 +104,71 @@ int32_t mz_stream_zstd_read(void *stream, void *buf, int32_t size) {
...
+    int32_t total_in = 0;
+    int32_t total_out = 0;
+    int32_t in_bytes = 0;
+    int32_t out_bytes = 0;
+    int32_t bytes_to_read = 0;
     int32_t read = 0;
-    size_t result = 0;
-
+    size_t errorCode = 0;
+    //int32_t err = LZMA_OK;
```
Comment (nmoinvaz) [inline mz_strm_zstd.c:117]: Remove comment?

File: mz_strm_zstd.c [hunk 6]
```diff
@@ -104,59 +104,71 @@ int32_t mz_stream_zstd_read(void *stream, void *buf, int32_t size) {
 #else
     mz_stream_zstd *zstd = (mz_stream_zstd *)stream;
     uint64_t total_in_before = 0;
-    uint64_t total_in_after = 0;
     uint64_t total_out_before = 0;
+    uint64_t total_in_after = 0;
```
Comment (nmoinvaz) [inline mz_strm_zstd.c:108]: `total_in_after` should go before..

**Comment by nmoinvaz** [comment]
Odd CI instances still failing on the "Run Tests". I think perhaps if you run `ctest . -C Debug` it will run all the tests and you can see where it is failing.

**Comment by fcharlie** [comment]
@nmoinvaz 

The failure of CI does not seem to be caused by zstd. Taking Windows GCC as an example, the failure here is caused by zlib:

```
C:/Program Files/PostgreSQL/12/lib/zlib.lib(crc32.obj):(.text$mn+0x1d7): undefined reference to `__security_check_cookie'
C:/Program Files/PostgreSQL/12/lib/zlib.lib(crc32.obj):(.xdata[$unwind$crc32_combine_]+0x18): undefined reference to `__GSHandlerCheck'
C:/Program Files/PostgreSQL/12/lib/zlib.lib(trees.obj):(.text$mn+0x13e): undefined reference to `__security_check_cookie'
C:/Program Files/PostgreSQL/12/lib/zlib.lib(trees.obj):(.xdata[$unwind$gen_codes]+0x8): undefined reference to `__GSHandlerCheck'
C:/Program Files/PostgreSQL/12/lib/zlib.lib(inftrees.obj):(.text$mn+0x161): undefined reference to `__security_check_cookie'
C:/Program Files/PostgreSQL/12/lib/zlib.lib(inftrees.obj):(.text$mn+0x50b): undefined reference to `__report_rangecheckfailure'
C:/Program Files/PostgreSQL/12/lib/zlib.lib(inftrees.obj):(.xdata[$unwind$inflate_table]+0x18): undefined reference to `__GSHandlerCheck'
collect2.exe: error: ld returned 1 exit status
```

**Comment by nmoinvaz** [comment]
The stalled CIs were due to it spinning in the loop. GCC is probably something totally unrelated. @gvollant fixes solved the problem.

**Comment by nmoinvaz** [comment]
@fcharlie thanks for all the code changes I know it was a pain.

**Comment by fcharlie** [comment]
@nmoinvaz I used the latest code from the minizip dev branch, and the decompression error seems to be back. Here is the `do-while` condition:

```c

    do {
        if (zstd->in.pos == zstd->in.size) {
            bytes_to_read = sizeof(zstd->buffer);
            if (zstd->max_total_in > 0) {
                if ((int64_t)bytes_to_read > (zstd->max_total_in - zstd->total_in))
                    bytes_to_read = (int32_t)(zstd->max_total_in - zstd->total_in);
            }

            read = mz_stream_read(zstd->stream.base, zstd->buffer, bytes_to_read);

            if (read < 0)
                return read;

            zstd->in.src = (const void*)zstd->buffer;
            zstd->in.pos = 0;
            zstd->in.size = (size_t)read;
        }

        total_in_before = zstd->in.pos;
        total_out_before = zstd->out.pos;

        result = ZSTD_decompressStream(zstd->zdstream, &zstd->out, &zstd->in);

        if (ZSTD_isError(result)) {
            zstd->error = (int32_t)result;
            return MZ_DATA_ERROR;
        }

        total_in_after = zstd->in.pos;
        total_out_after = zstd->out.pos;
        if ((zstd->max_total_out != -1) && (int64_t)total_out_after > zstd->max_total_out)
            total_out_after = (uint64_t)zstd->max_total_out;

        in_bytes = (int32_t)(total_in_after - total_in_before);
        out_bytes = (int32_t)(total_out_after - total_out_before);

        total_in += in_bytes;
        total_out += out_bytes;

        zstd->total_in += in_bytes;
        zstd->total_out += out_bytes;

    } while (zstd->in.pos < zstd->in.size && zstd->out.pos < zstd->out.size);

```
When ZSTD_DStream does not have enough data (it does not seem to satisfy a decompression window), ZSTD_inBuffer has been copied to ZSTD_DStream, but there is no data in ZSTD_outBuffer, then the condition judgment will appear the following situation:

```c
zstd->in.pos == zstd->in.size
```
The result is that mz_stream_zstd_read returns zero. According to the logic of minizip, the stream has been parsed, but the size of uncompressed is wrong, so the decompression fails.

Please point out if there is anything wrong with what I said, thank you.

**Comment by gvollant** [comment]
Yann Collet @Cyan4973 suggested me this condition (on another projet, but similar problem)
````
while( in.pos < in.size || out.pos == out.size )
````

**Comment by nmoinvaz** [comment]
@fcharlie can you verify that latest `dev` branch solves the issue?

**Comment by fcharlie** [comment]
@nmoinvaz I looked at the dev branch and the problem has been fixed

**Pull request merged**